question28/js/forgotpassword.js

```diff
@@ -0,0 +1,68 @@
+'use strict';
+
+import { validationInfo } from "./validation";
+import { Chance } from "chance";
+const chance = new Chance();
+
+const submitButton = document.getElementById('js-submit-button');
+const userInfoInputArea = document.getElementById('name_mail');
+const inputArea = document.getElementById('name_mail');
+
+const isValidStatus = {
+  name_mail: false,
+}
+
+const setValidationEvents = (isValid, targetForm) =>
+  isValid ? validEvent(targetForm) : invalidEvent(targetForm);
+
+const validEvent = (targetForm) => {
+  targetForm.classList.add("valid");
+  targetForm.classList.remove("invalid");
+  targetForm.nextElementSibling.textContent = "";
+  isValidStatus[targetForm.id] = true;
+};
+
+const invalidEvent = (targetForm) => {
+  targetForm.classList.add("invalid");
+  targetForm.classList.remove("valid");
+  targetForm.nextElementSibling.textContent = validationInfo[targetForm.id].errorMessage;
+  isValidStatus[targetForm.id] = false;
+};
+
+const validateInputValue = (e) => {
+  const targetForm = e.target;
+  const value = targetForm.value.trim();
+  const result = validationInfo[targetForm.id].validation(value);
+  setValidationEvents(result,targetForm);
+  toggleDisabledOfSubmitButton(checkAllValidity());
+} 
+
+const checkAllValidity = () => {
+  return Object.values(isValidStatus).every((result) => result );
+}
+
+const toggleDisabledOfSubmitButton = (isValid) => {
+  submitButton.disabled = isValid ? false : true;
+} 
```

Comment:
Thanks for pointing that out!
Since the login page, register page, and forgotpassword page used the same function for validation, I divided the common part into validation.js. But there are some parts that are connected to others, and it is no longer a simple fix...I know this is going to be a tough review, so if you notice anything, I'd be glad to know 🙇 

fixed commit : https://github.com/csakyo/homework/pull/42/commits/d9ef8de52fe71ba88c2841ed920cbef0b139f077

Reply:
修正ありがとうございます！日本語ですみません。

`isValidStatus` で、バリデーションの状態管理をされているかと思うのですが、
`isValidStatus` の状態を変更するために、
変数`pageType`が実際には使われない関数の中にも渡されているのがやや引っかかりました。

バリデーションチェックにひっかかったものには、invalidクラスを付与しているかと思うので、
invalidクラスが一つもない場合は、checkAllValidityをtrueにすれば、
pageTypeをバケツリレーすることなく、すべてのページでバリデーションを共通化できると思いました。

パフォーマンス的に、オブジェクトで状態管理したほうがいいのかもしれないので、
もし何か問題がありそうならご指摘いただけると幸いです。

下記に、コードを示します。
[CodeSandBox](https://codesandbox.io/s/infallible-sun-oyjdk9?file=/js/validations.js)

Reply:
まいさん、細かく見ていただきまして心から感謝です:bow:
参考コードありがとうございました！！
ご指摘箇所自分もどうかな..となっていたので、まいさんのアイディアを参考にさせていただきました:bow:

「invalidクラスが一つもない場合」ですが、最初の時点ではvalidもinvalidも設定していないので、invalidクラスが一つもない状況であり、最初のblurイベントでバリデーションが通ってしまっています。

https://user-images.githubusercontent.com/51282804/204128383-0a44c54f-f66d-4e1d-a85c-790a308dfb2b.mov

よって、逆の発想でinputタグの数とvalidクラスの数が一致した場合というロジックで組んでみたのですがどうでしょう...また、さえさんにご指摘いただいたのですが、「モジュールは再利用可能で自己完結型であることが望ましい」とのことで、モジュールでの値の取得をするべきではないと教わりました。よって、値取得が必要な関数は一旦モジュールから外しました。

お忙しいところ、重たいレビューになってしまい申し訳ありません...:bow:
ご確認いただけますと幸いです。よろしくお願いいたします。

Fixed commit :https://github.com/csakyo/homework/pull/42/commits/03e50c027820f69fdbf6d1a864c84a759c5db67f

Reply:
こちらの指摘ありがとうございます！
確かにその通りです。
> 「invalidクラスが一つもない場合」ですが、最初の時点ではvalidもinvalidも設定していないので、invalidクラスが一つもない状況であり、最初のblurイベントでバリデーションが通ってしまっています。

下記、良い解決策だと感じました。
自分は、最初に付与してから取り除くことを考えたので、違う方法を教えていただきありがとうございます！
> よって、逆の発想でinputタグの数とvalidクラスの数が一致した場合というロジックで組んでみたのですがどうでしょう...

こちらについては、勉強不足ですみませんでした:sweat_drops:
承知しました！
> また、さえさんにご指摘いただいたのですが、「モジュールは再利用可能で自己完結型であることが望ましい」とのことで、モジュールでの値の取得をするべきではないと教わりました。よって、値取得が必要な関数は一旦モジュールから外しました。